### PR TITLE
Added the PIN APIs

### LINKED ISSUE #9: Add pin APIs
Add all crud APIs for Pin
1. Add new pin data
2. Edit pin
3. Delete pin
4. Get pin by id/query

### PR DESCRIPTION
**Pull request by haseebzaki-07**
fixes #9 

Successfully added all the operations for PIN with optimal codes and better error logging. Also edited the database connecting logic for better error logging and messaging.

Create PIN route test:

Screenshot 2024-10-10 225809

Update PIN route test:

Screenshot 2024-10-10 225945

Get PIN route test:

Screenshot 2024-10-10 225848

Delete PIN route test:

Screenshot 2024-10-10 225959

### CODE REVIEW

**Comment by haseebzaki-07** [comment]
@Mustafiz04  The update and delete routes were not working I fixed them and also referenced the pins with the user generating them for PIN management. Maintained authentication across all routes and optimized the codes.

File: api/controller/pin.js [hunk 1]
```diff
@@ -1,88 +1,80 @@
...
+import { createPin, deletePin, getPinById, updatePin } from "../model";
 import User from "../model/user";
 
 const addNewPin = async (req, res) => {
-  const { body = {} } = req;
+  const { body } = req;
+
+  if (!body.latitude || !body.longitude || !body.title) {
+    return res.status(400).json({ message: "Latitude, longitude, and title are required." });
+  }
```
Comment (Mustafiz04) [inline api/controller/pin.js:9]: Can you do this validation using express-validator middleware ??

File: api/controller/pin.js [hunk 2]
```diff
@@ -1,88 +1,80 @@
...
-        errors[key] = error.errors[key].message;
-        return errors;
-      }, {});
-    } else if (name === "MongoError" && code === 11000) {
-      status = 409;
-      error_message = "Email already exists";
-    }
-    return res.status(status).json({ errors, message: error_message });
+    console.error(error);
+    return res.status(500).json({ message: "Failed to create pin", error: error.message });
```
Comment (Mustafiz04) [inline api/controller/pin.js:18]: What about other possible errors? This way it will only give 500 error code

File: api/controller/pin.js [hunk 3]
```diff
@@ -1,88 +1,80 @@
...
+
   try {
     const user = req.user;
-    const updatedPin = await updatePin(id, body);
-    if (!updatedPin) {
-      return res.status(404).json({ error: "Pin not found" });
+    const pin = await getPinById(id);
+    
+    if (!pin) {
+      return res.status(404).json({ message: "Pin not found" });
```
Comment (Mustafiz04) [inline api/controller/pin.js:30]: Same using express-validator

**Review by Mustafiz04 — COMMENTED** [review]
Code looks excellent. Just need to polish and structure the code.
1. Use validation middleware, which will check there and throw all possible errors rather than checking in the controller.
2. Error handling.

Do the same for the other PR once this one is merged.

File: api/controller/pin.js (hunk 1 above)
Comment (haseebzaki-07) [reply api/controller/pin.js:9]: sure I will try it

**Comment by haseebzaki-07** [comment]
Hey @Mustafiz04 I added express-validator middlewares and also better error logging in the codes.

**Comment by haseebzaki-07** [comment]
Hey @Mustafiz04 please finalize this PR so that I can make changes in the other one #24

**Comment by haseebzaki-07** [comment]
hey @Mustafiz04 please review this PR

File: api/middleware/auth.js [hunk 4]
```diff
@@ -1,39 +1,58 @@
...
+
+export const validateUpdatePin = [
+  param('id').isMongoId().withMessage('Invalid pin ID format'),
+
+  body('latitude').optional().isFloat({ min: -90, max: 90 }).withMessage('Latitude must be between -90 and 90'),
+  body('longitude').optional().isFloat({ min: -180, max: 180 }).withMessage('Longitude must be between -180 and 180'),
+  body('title').optional().isString().isLength({ min: 3 }).withMessage('Title must be at least 3 characters long'),
+  body('description').optional().isString().isLength({ max: 500 }).withMessage('Description cannot be longer than 500 characters'),
+
+];
```
Comment (Mustafiz04) [inline api/middleware/auth.js:58]: We can move this to new middleware file `pin.js` so we can have all the pin related middleware login in one place.

File: api/controller/pin.js [hunk 5]
```diff
@@ -1,88 +1,146 @@
...
+import { createPin, deletePin, getPinById, updatePin } from "../model";
 import User from "../model/user";
 
 const addNewPin = async (req, res) => {
-  const { body = {} } = req;
+  const errors = validationResult(req);
+  if (!errors.isEmpty()) {
+    console.error("Validation errors:", errors.array());
+    return res.status(400).json({ errors: errors.array() });
+  }
```
Comment (Mustafiz04) [inline api/controller/pin.js:10]: This you can move to middleware only. No need to write in each controller. Our motive is to handle these there only.

**Review by Mustafiz04 — CHANGES_REQUESTED** [review]
These two changes, rest look good.

File: api/middleware/auth.js (hunk 4 above)
Comment (haseebzaki-07) [reply api/middleware/auth.js]: Done

File: api/controller/pin.js (hunk 5 above)
Comment (haseebzaki-07) [reply api/controller/pin.js]: Done

**Comment by haseebzaki-07** [comment]
hey @Mustafiz04 I made a new Pin.js middleware and relocated the desired codes there and updated the routes. Could you please increase the level of the PR as fine tuning this thing was bit of a task.

**Comment by haseebzaki-07** [comment]
hey @Mustafiz04 please finalize this PR

**Comment by Mustafiz04** [comment]
Got this error while deploying. Please check

Screenshot 2024-10-22 at 11 13 54 PM

**Comment by haseebzaki-07** [comment]
sure

**Pull request merged**